File: swift/BCEnvelope/Tests/BCEnvelopeTests/ObscuringTests.swift
import Testing
import BCComponents
import BCEnvelope
import WolfBase
import Foundation

struct ObscuringTests {
    /// This tests the transformation of different kinds of "obscured" envelopes into
    /// others. Some transformations are allowed, some are idempotent (return the same
    /// result), and some throw errors.
    ///
    /// | Operation > | Encrypt | Elide      | Compress   |
    /// |:------------|:--------|:-----------|:-----------|
    /// | Encrypted   | ERROR   | OK         | ERROR      |
    /// | Elided      | ERROR   | IDEMPOTENT | ERROR      |
    /// | Compressed  | OK      | OK         | IDEMPOTENT |
    ///
    @Test func testObscuring() throws {
        let key = SymmetricKey()
        
        let envelope = Envelope(plaintextHello)
        #expect(!envelope.isObscured)
        
        let encrypted = try envelope.encryptSubject(with: key)
        #expect(encrypted.isObscured)

        let elided = envelope.elide()
        #expect(elided.isObscured)

        let compressed = try envelope.compress()
        #expect(compressed.isObscured)

        
        // ENCRYPTION
        
        // Cannot encrypt an encrypted envelope.
        //
        // If allowed, would result in an envelope with the same digest but
        // double-encrypted, possibly with a different key, which is probably not what's
        // intended. If you want to double-encrypt then wrap the encrypted envelope first,
        // which will change its digest.
        #expect(throws: (any Swift.Error).self) { try encrypted.encryptSubject(with: key) }
        
        // Cannot encrypt an elided envelope.
        //
        // Elided envelopes have no data to encrypt.
        #expect(throws: (any Swift.Error).self) { try elided.encryptSubject(with: key) }
        
        // OK to encrypt a compressed envelope.
        guard case .encrypted = try compressed.encryptSubject(with: key) else {
            Issue.record()
            return
        }
        
        
        // ELISION
        
        // OK to elide an encrypted envelope.
        guard case .elided = encrypted.elide() else {
            Issue.record()
            return
        }
        
        // Eliding an elided envelope is idempotent.
        guard case .elided = elided.elide() else {
            Issue.record()
            return
        }
        
        // OK to elide a compressed envelope.
        guard case .elided = compressed.elide() else {
            Issue.record()
            return
        }
        
        
        // COMPRESSION
        
        // Cannot compress an encrypted envelope.
        //
        // Encrypted envelopes cannot become smaller because encrypted data looks random,
        // and random data is not compressible.
        #expect(throws: (any Swift.Error).self) { try encrypted.compress() }
        
        // Cannot compress an elided envelope.
        //
        // Elided envelopes have no data to compress.
        #expect(throws: (any Swift.Error).self) { try elided.compress() }
        
        // Compressing a compressed envelope is idempotent.
        guard case .compressed = try compressed.compress() else {
            Issue.record()
            return
        }
    }
    
    @Test func testNodesMatching() throws {
        let envelope = Envelope("Alice")
            .addAssertion("knows", "Bob")
            .addAssertion("age", 30)
            .addAssertion("city", "Boston")
        
        let knowsAssertion = try envelope.assertion(withPredicate: "knows")
        let knowsDigest = knowsAssertion.digest
        
        let ageAssertion = try envelope.assertion(withPredicate: "age")
        let ageDigest = ageAssertion.digest
        
        let elideTarget: Swift.Set<Digest> = [knowsDigest]
        let compressTarget: Swift.Set<Digest> = [ageDigest]
        
        var obscured = envelope.elideRemoving(elideTarget)
        obscured = obscured.elideRemoving(compressTarget, action: .compress)
        
        #expect(obscured.format() ==
        """
        "Alice" [
            "city": "Boston"
            COMPRESSED
            ELIDED
        ]
        """
        )
        
        let elidedNodes = obscured.nodesMatching(nil, [.elided])
        #expect(elidedNodes.contains(knowsDigest))
        
        let compressedNodes = obscured.nodesMatching(nil, [.compressed])
        #expect(compressedNodes.contains(ageDigest))
        
        let targetFilter: Swift.Set<Digest> = [knowsDigest]
        let filtered = obscured.nodesMatching(targetFilter, [.elided])
        #expect(filtered.count == 1)
        #expect(filtered.contains(knowsDigest))
        
        let allInTarget = obscured.nodesMatching(elideTarget, [])
        #expect(allInTarget.count == 1)
        #expect(allInTarget.contains(knowsDigest))
        
        let noMatchTarget: Swift.Set<Digest> = [Digest.fromImage(Data("nonexistent".utf8))]
        let noMatches = obscured.nodesMatching(noMatchTarget, [.elided])
        #expect(noMatches.isEmpty)
    }
    
    @Test func testWalkUnelide() throws {
        let alice = Envelope("Alice")
        let bob = Envelope("Bob")
        let carol = Envelope("Carol")
        
        let envelope = Envelope("Alice")
            .addAssertion("knows", "Bob")
            .addAssertion("friend", "Carol")
        
        let elided = envelope
            .elideRemoving(alice)
            .elideRemoving(bob)
        
        #expect(elided.format() ==
        """
        ELIDED [
            "friend": "Carol"
            "knows": ELIDED
        ]
        """
        )
        
        let restored = elided.walkUnelide([alice, bob, carol])
        
        #expect(restored.format() ==
        """
        "Alice" [
            "friend": "Carol"
            "knows": "Bob"
        ]
        """
        )
        
        let partial = elided.walkUnelide([alice])
        #expect(partial.format() ==
        """
        "Alice" [
            "friend": "Carol"
            "knows": ELIDED
        ]
        """
        )
        
        let unchanged = elided.walkUnelide([])
        #expect(unchanged.isIdentical(to: elided))
    }
    
    @Test func testWalkDecrypt() throws {
        let key1 = SymmetricKey()
        let key2 = SymmetricKey()
        let key3 = SymmetricKey()
        
        let envelope = Envelope("Alice")
            .addAssertion("knows", "Bob")
            .addAssertion("age", 30)
            .addAssertion("city", "Boston")
        
        let knowsAssertion = try envelope.assertion(withPredicate: "knows")
        let ageAssertion = try envelope.assertion(withPredicate: "age")
        
        let encrypt1Target: Swift.Set<Digest> = [knowsAssertion.digest]
        let encrypt2Target: Swift.Set<Digest> = [ageAssertion.digest]
        
        let encrypted = envelope
            .elideRemoving(encrypt1Target, action: .encrypt(key1))
            .elideRemoving(encrypt2Target, action: .encrypt(key2))
        
        #expect(encrypted.format() ==
        """
        "Alice" [
            "city": "Boston"
            ENCRYPTED (2)
        ]
        """
        )
        
        let decrypted = encrypted.walkDecrypt([key1, key2])
        #expect(decrypted.format() ==
        """
        "Alice" [
            "age": 30
            "city": "Boston"
            "knows": "Bob"
        ]
        """
        )
        
        let partial = encrypted.walkDecrypt([key1])
        #expect(!partial.isIdentical(to: encrypted))
        #expect(partial.isEquivalent(to: envelope))
        #expect(partial.format() ==
        """
        "Alice" [
            "city": "Boston"
            "knows": "Bob"
            ENCRYPTED
        ]
        """
        )
        
        let unchanged = encrypted.walkDecrypt([key3])
        #expect(unchanged.isIdentical(to: encrypted))
    }
    
    @Test func testWalkDecompress() throws {
        let envelope = Envelope("Alice")
            .addAssertion("knows", "Bob")
            .addAssertion("bio", String(repeating: "A", count: 1000))
            .addAssertion("description", String(repeating: "B", count: 1000))
        
        let bioAssertion = try envelope.assertion(withPredicate: "bio")
        let descriptionAssertion = try envelope.assertion(withPredicate: "description")
        
        let bioDigest = bioAssertion.digest
        let descriptionDigest = descriptionAssertion.digest
        
        let compressTarget: Swift.Set<Digest> = [bioDigest, descriptionDigest]
        let compressed = envelope.elideRemoving(compressTarget, action: .compress)
        
        #expect(compressed.format() ==
        """
        "Alice" [
            "knows": "Bob"
            COMPRESSED (2)
        ]
        """
        )
        
        let decompressed = compressed.walkDecompress(nil)
        #expect(decompressed.isEquivalent(to: envelope))
        
        let target: Swift.Set<Digest> = [bioDigest]
        let partial = compressed.walkDecompress(target)
        #expect(!partial.isIdentical(to: compressed))
        #expect(partial.isEquivalent(to: envelope))
        
        let stillCompressed = partial.nodesMatching(nil, [.compressed])
        #expect(stillCompressed.contains(descriptionDigest))
        #expect(!stillCompressed.contains(bioDigest))
        
        let noMatch: Swift.Set<Digest> = [Digest.fromImage(Data("nonexistent".utf8))]
        let unchanged = compressed.walkDecompress(noMatch)
        #expect(unchanged.isIdentical(to: compressed))
    }
    
    @Test func testMixedObscurationOperations() throws {
        let key = SymmetricKey()
        
        let envelope = Envelope("Alice")
            .addAssertion("knows", "Bob")
            .addAssertion("age", 30)
            .addAssertion("bio", String(repeating: "A", count: 1000))
        
        let knowsAssertion = try envelope.assertion(withPredicate: "knows")
        let ageAssertion = try envelope.assertion(withPredicate: "age")
        let bioAssertion = try envelope.assertion(withPredicate: "bio")
        
        let knowsDigest = knowsAssertion.digest
        let ageDigest = ageAssertion.digest
        let bioDigest = bioAssertion.digest
        
        let elideTarget: Swift.Set<Digest> = [knowsDigest]
        let encryptTarget: Swift.Set<Digest> = [ageDigest]
        let compressTarget: Swift.Set<Digest> = [bioDigest]
        
        let obscured = envelope
            .elideRemoving(elideTarget)
            .elideRemoving(encryptTarget, action: .encrypt(key))
            .elideRemoving(compressTarget, action: .compress)
        
        let elided = obscured.nodesMatching(nil, [.elided])
        let encrypted = obscured.nodesMatching(nil, [.encrypted])
        let compressed = obscured.nodesMatching(nil, [.compressed])
        
        #expect(elided.contains(knowsDigest))
        #expect(encrypted.contains(ageDigest))
        #expect(compressed.contains(bioDigest))
        
        let restored = obscured
            .walkUnelide([knowsAssertion])
            .walkDecrypt([key])
            .walkDecompress(nil)
        
        #expect(restored.isEquivalent(to: envelope))
    }
}

File: csharp/BCEnvelope/BCEnvelope.Tests/ObscuringTests.cs
using BlockchainCommons.BCComponents;
using BlockchainCommons.BCEnvelope;

namespace BlockchainCommons.BCEnvelope.Tests;

public sealed class ObscuringTests
{
    private const string PlaintextHello = "Hello.";

    [Fact]
    public void TestObscuring()
    {
        var key = SymmetricKey.New();

        var envelope = Envelope.Create(PlaintextHello);
        Assert.False(envelope.IsObscured);

        var encrypted = envelope.EncryptSubject(key);
        Assert.True(encrypted.IsObscured);

        var elided = envelope.Elide();
        Assert.True(elided.IsObscured);

        var compressed = envelope.Compress();
        Assert.True(compressed.IsObscured);

        // ENCRYPTION

        // Cannot encrypt an encrypted envelope.
        Assert.ThrowsAny<Exception>(() => encrypted.EncryptSubject(key));

        // Cannot encrypt an elided envelope.
        Assert.ThrowsAny<Exception>(() => elided.EncryptSubject(key));

        // OK to encrypt a compressed envelope.
        var encryptedCompressed = compressed.EncryptSubject(key);
        Assert.True(encryptedCompressed.IsEncrypted);

        // ELISION

        // OK to elide an encrypted envelope.
        var elidedEncrypted = encrypted.Elide();
        Assert.True(elidedEncrypted.IsElided);

        // Eliding an elided envelope is idempotent.
        var elidedElided = elided.Elide();
        Assert.True(elidedElided.IsElided);

        // OK to elide a compressed envelope.
        var elidedCompressed = compressed.Elide();
        Assert.True(elidedCompressed.IsElided);

        // COMPRESSION

        // Cannot compress an encrypted envelope.
        Assert.ThrowsAny<Exception>(() => encrypted.Compress());

        // Cannot compress an elided envelope.
        Assert.ThrowsAny<Exception>(() => elided.Compress());

        // Compressing a compressed envelope is idempotent.
        var compressedCompressed = compressed.Compress();
        Assert.True(compressedCompressed.IsCompressed);
    }

    [Fact]
    public void TestNodesMatching()
    {
        var envelope = Envelope.Create("Alice")
            .AddAssertion("knows", "Bob")
            .AddAssertion("age", 30)
            .AddAssertion("city", "Boston");

        // Get some digests for targeting
        var knowsAssertion = envelope.AssertionWithPredicate("knows");
        var knowsDigest = knowsAssertion.GetDigest();

        var ageAssertion = envelope.AssertionWithPredicate("age");
        var ageDigest = ageAssertion.GetDigest();

        // Elide one assertion, compress another
        var elideTarget = new HashSet<Digest> { knowsDigest };

        var compressTarget = new HashSet<Digest> { ageDigest };

        var obscured = envelope.ElideRemovingSet(elideTarget);

        obscured = obscured.ElideRemovingSetWithAction(
            compressTarget,
            ObscureAction.Compress);

        // Verify the structure with elided and compressed nodes
        Assert.Equal(
            "\"Alice\" [\n" +
            "    \"city\": \"Boston\"\n" +
            "    COMPRESSED\n" +
            "    ELIDED\n" +
            "]",
            obscured.Format());

        // Test finding elided nodes
        var elidedNodes = obscured.NodesMatching(null, ObscureType.Elided);
        Assert.Contains(knowsDigest, elidedNodes);

        // Test finding compressed nodes
        var compressedNodes = obscured.NodesMatching(null, ObscureType.Compressed);
        Assert.Contains(ageDigest, compressedNodes);

        // Test finding with target filter
        var targetFilter = new HashSet<Digest> { knowsDigest };
        var filtered = obscured.NodesMatching(targetFilter, ObscureType.Elided);
        Assert.Single(filtered);
        Assert.Contains(knowsDigest, filtered);

        // Test finding all obscured nodes (no type filter)
        var allInTarget = obscured.NodesMatching(elideTarget);
        Assert.Single(allInTarget);
        Assert.Contains(knowsDigest, allInTarget);

        // Test with no matches
        var noMatchTarget = new HashSet<Digest> { Digest.FromImage("nonexistent"u8.ToArray()) };
        var noMatches = obscured.NodesMatching(noMatchTarget, ObscureType.Elided);
        Assert.Empty(noMatches);
    }

    [Fact]
    public void TestWalkUnelide()
    {
        var alice = Envelope.Create("Alice");
        var bob = Envelope.Create("Bob");
        var carol = Envelope.Create("Carol");

        var envelope = Envelope.Create("Alice")
            .AddAssertion("knows", "Bob")
            .AddAssertion("friend", "Carol");

        // Elide multiple parts
        var elided = envelope
            .ElideRemovingTarget(alice)
            .ElideRemovingTarget(bob);

        // Verify parts are elided
        Assert.Equal(
            "ELIDED [\n" +
            "    \"friend\": \"Carol\"\n" +
            "    \"knows\": ELIDED\n" +
            "]",
            elided.Format());

        // Restore with walk_unelide
        var restored = elided.WalkUnelide(new[] { alice, bob, carol });

        // The restored envelope should match original
        Assert.Equal(
            "\"Alice\" [\n" +
            "    \"friend\": \"Carol\"\n" +
            "    \"knows\": \"Bob\"\n" +
            "]",
            restored.Format());

        // Test with partial restoration (only some envelopes provided)
        var partial = elided.WalkUnelide(new[] { alice });
        Assert.Equal(
            "\"Alice\" [\n" +
            "    \"friend\": \"Carol\"\n" +
            "    \"knows\": ELIDED\n" +
            "]",
            partial.Format());

        // Test with no matching envelopes
        var unchanged = elided.WalkUnelide(Array.Empty<Envelope>());
        Assert.True(unchanged.IsIdenticalTo(elided));
    }

    [Fact]
    public void TestWalkDecrypt()
    {
        var key1 = SymmetricKey.New();
        var key2 = SymmetricKey.New();
        var key3 = SymmetricKey.New();

        var envelope = Envelope.Create("Alice")
            .AddAssertion("knows", "Bob")
            .AddAssertion("age", 30)
            .AddAssertion("city", "Boston");

        // Encrypt different parts with different keys
        var knowsAssertion = envelope.AssertionWithPredicate("knows");
        var ageAssertion = envelope.AssertionWithPredicate("age");

        var encrypt1Target = new HashSet<Digest> { knowsAssertion.GetDigest() };
        var encrypt2Target = new HashSet<Digest> { ageAssertion.GetDigest() };

        var encrypted = envelope
            .ElideRemovingSetWithAction(
                encrypt1Target,
                ObscureAction.Encrypt(key1))
            .ElideRemovingSetWithAction(
                encrypt2Target,
                ObscureAction.Encrypt(key2));

        // Verify parts are encrypted
        Assert.Equal(
            "\"Alice\" [\n" +
            "    \"city\": \"Boston\"\n" +
            "    ENCRYPTED (2)\n" +
            "]",
            encrypted.Format());

        // Decrypt with all keys
        var decrypted = encrypted.WalkDecrypt(new[] { key1, key2 });
        Assert.Equal(
            "\"Alice\" [\n" +
            "    \"age\": 30\n" +
            "    \"city\": \"Boston\"\n" +
            "    \"knows\": \"Bob\"\n" +
            "]",
            decrypted.Format());

        // Decrypt with only one key (partial decryption)
        var partial = encrypted.WalkDecrypt(new[] { key1 });
        Assert.False(partial.IsIdenticalTo(encrypted));
        // Note: partial is still equivalent because encrypted nodes preserve
        // digests
        Assert.True(partial.IsEquivalentTo(envelope));

        // There should still be one encrypted node remaining
        Assert.Equal(
            "\"Alice\" [\n" +
            "    \"city\": \"Boston\"\n" +
            "    \"knows\": \"Bob\"\n" +
            "    ENCRYPTED\n" +
            "]",
            partial.Format());

        // Decrypt with wrong key (should be unchanged)
        var unchanged = encrypted.WalkDecrypt(new[] { key3 });
        Assert.True(unchanged.IsIdenticalTo(encrypted));
    }

    [Fact]
    public void TestWalkDecompress()
    {
        var envelope = Envelope.Create("Alice")
            .AddAssertion("knows", "Bob")
            .AddAssertion("bio", new string('A', 1000))
            .AddAssertion("description", new string('B', 1000));

        // Compress multiple parts
        var bioAssertion = envelope.AssertionWithPredicate("bio");
        var descAssertion = envelope.AssertionWithPredicate("description");

        var bioDigest = bioAssertion.GetDigest();
        var descDigest = descAssertion.GetDigest();

        var compressTarget = new HashSet<Digest> { bioDigest, descDigest };

        var compressed = envelope.ElideRemovingSetWithAction(
            compressTarget,
            ObscureAction.Compress);

        // Verify parts are compressed
        Assert.Equal(
            "\"Alice\" [\n" +
            "    \"knows\": \"Bob\"\n" +
            "    COMPRESSED (2)\n" +
            "]",
            compressed.Format());

        // decompress all
        var decompressed = compressed.WalkDecompress();
        Assert.True(decompressed.IsEquivalentTo(envelope));

        // Decompress with target filter (only one node)
        var target = new HashSet<Digest> { bioDigest };

        var partialDecomp = compressed.WalkDecompress(target);
        Assert.False(partialDecomp.IsIdenticalTo(compressed));
        // Note: partial is still equivalent because compressed nodes preserve
        // digests
        Assert.True(partialDecomp.IsEquivalentTo(envelope));

        // Bio should be decompressed but description still compressed
        var stillCompressed = partialDecomp.NodesMatching(null, ObscureType.Compressed);
        Assert.Contains(descDigest, stillCompressed);
        Assert.DoesNotContain(bioDigest, stillCompressed);

        // Decompress with non-matching target (should be unchanged)
        var noMatch = new HashSet<Digest> { Digest.FromImage("nonexistent"u8.ToArray()) };
        var unchangedDecomp = compressed.WalkDecompress(noMatch);
        Assert.True(unchangedDecomp.IsIdenticalTo(compressed));
    }

    [Fact]
    public void TestMixedObscurationOperations()
    {
        var key = SymmetricKey.New();

        var envelope = Envelope.Create("Alice")
            .AddAssertion("knows", "Bob")
            .AddAssertion("age", 30)
            .AddAssertion("bio", new string('A', 1000));

        var knowsAssertion = envelope.AssertionWithPredicate("knows");
        var ageAssertion = envelope.AssertionWithPredicate("age");
        var bioAssertion = envelope.AssertionWithPredicate("bio");

        var knowsDigest = knowsAssertion.GetDigest();
        var ageDigest = ageAssertion.GetDigest();
        var bioDigest = bioAssertion.GetDigest();

        // Apply different obscuration types
        var elideTarget = new HashSet<Digest> { knowsDigest };
        var encryptTarget = new HashSet<Digest> { ageDigest };
        var compressTarget = new HashSet<Digest> { bioDigest };

        var obscured = envelope
            .ElideRemovingSet(elideTarget)
            .ElideRemovingSetWithAction(
                encryptTarget,
                ObscureAction.Encrypt(key))
            .ElideRemovingSetWithAction(
                compressTarget,
                ObscureAction.Compress);

        // Verify different obscuration types
        var elidedNodes = obscured.NodesMatching(null, ObscureType.Elided);
        var encryptedNodes = obscured.NodesMatching(null, ObscureType.Encrypted);
        var compressedNodes = obscured.NodesMatching(null, ObscureType.Compressed);

        Assert.Contains(knowsDigest, elidedNodes);
        Assert.Contains(ageDigest, encryptedNodes);
        Assert.Contains(bioDigest, compressedNodes);

        // Restore everything
        var restored = obscured
            .WalkUnelide(new[] { knowsAssertion })
            .WalkDecrypt(new[] { key })
            .WalkDecompress();

        Assert.True(restored.IsEquivalentTo(envelope));
    }
}
